addons/hearing/functions/fnc_addEarPlugs.sqf

```diff
@@ -15,53 +15,85 @@
  * Public: No
  */
 
-// only run this after the settings are initialized
-if !(EGVAR(common,settingsInitFinished)) exitWith {
+// Only run this after the settings are initialized
+if (!EGVAR(common,settingsInitFinished)) exitWith {
     EGVAR(common,runAtSettingsInitialized) pushBack [FUNC(addEarPlugs), _this];
 };
 
+// Exit if hearing is disabled or if autoAdd is disabled
+if (!GVAR(enableCombatDeafness) || {GVAR(autoAddEarplugsToUnits) == 0}) exitWith {};
+
 params ["_unit"];
 TRACE_2("params",_unit,typeOf _unit);
 
-// Exit if hearing is disabled OR autoAdd is disabled OR soldier has earplugs already in (persistence scenarios)
-if (!GVAR(enableCombatDeafness) || {GVAR(autoAddEarplugsToUnits) == 0} || {[_unit] call FUNC(hasEarPlugsIn)}) exitWith {};
+// Exit if soldier already has earplugs (in ears (persistence scenarios) or inventory)
+if (_unit call FUNC(hasEarPlugsIn) || {[_unit, "ACE_EarPlugs"] call EFUNC(common,hasItem)}) exitWith {};
 
 // Add earplugs if enabled for everyone or if the soldier has a rocket launcher
 if (GVAR(autoAddEarplugsToUnits) == 2 || {(secondaryWeapon _unit) != ""}) exitWith {
     TRACE_1("has launcher - adding",_unit);
     _unit addItem "ACE_EarPlugs";
 };
 
-// otherwise add earplugs if the soldier has a big rifle
-if ((primaryWeapon _unit) == "") exitWith {};
+// Otherwise add earplugs if the soldier has a big rifle
+private _weapon = primaryWeapon _unit;
 
-(primaryWeaponMagazine _unit) params [["_magazine", ""]];
-if (_magazine == "") exitWith {};
+if (_weapon == "") exitWith {};
 
-private _cfgMagazine = configFile >> "CfgMagazines" >> _magazine;
+// Cache maximum loadness for future calls
+private _maxLoudness = GVAR(cacheMaxAmmoLoudness) getOrDefaultCall [_weapon, {
+    private _cfgMagazines = configFile >> "CfgMagazines";
 
-private _initSpeed = getNumber (_cfgMagazine >> "initSpeed");
-private _ammo = getText (_cfgMagazine >> "ammo");
-private _count = getNumber (_cfgMagazine >> "count");
+    // Get the weapon's compatible magazines, so that all magazines are cached
+    private _maxLoudness = selectMax ((compatibleMagazines _weapon) apply {
+        private _magazine = _x;
+        private _cfgMagazine = _cfgMagazines >> _magazine;
+        private _ammo = getText (_cfgMagazine >> "ammo");
 
-private _cfgAmmo = configFile >> "CfgAmmo";
+        GVAR(cacheAmmoLoudness) getOrDefaultCall [_magazine, {
+            private _cfgAmmo = configFile >> "CfgAmmo";
+            private _initSpeed = getNumber (configFile >> "CfgMagazines" >> _magazine >> "initSpeed");
```

Comment:
```suggestion
            private _initSpeed = getNumber (_cfgMagazine >> "initSpeed");
```
but I understand not doing this so firedNear is the exact same code base

Reply:
If anything, it would be wise to make it a function, to avoid duplicate code. I'll do that.